### ISSUE TITLE
flags with length > 1 cause errors when printing `training_run()` output

### ISSUE THREAD

**Issue by kevinushey** [rstudio/tfruns#15]
https://github.com/rstudio/cloudml/blob/master/examples/census/train.R

The run succeeds, but the format of the returned `viewed_run` data is off:

```R
> str(value)
Classes 'tfruns_viewed_run', 'tfruns_runs_df', 'tbl_df', 'tbl' and 'data.frame':	4 obs. of  24 variables:
 $ run_dir                      : chr  "runs/2017-09-27T19-49-07Z" "runs/2017-09-27T19-49-07Z" "runs/2017-09-27T19-49-07Z" "runs/2017-09-27T19-49-07Z"
 $ metric_mean_losses           : num  19.9 19.9 19.9 19.9
 $ metric_total_losses          : num  19.9 19.9 19.9 19.9
 $ flag_train_file              : chr  "gs://cloudml-public/census/data/adult.data.csv" "gs://cloudml-public/census/data/adult.data.csv" "gs://cloudml-public/census/data/adult.data.csv" "gs://cloudml-public/census/data/adult.data.csv"
 $ flag_eval_file               : chr  "gs://cloudml-public/census/data/adult.test.csv" "gs://cloudml-public/census/data/adult.test.csv" "gs://cloudml-public/census/data/adult.test.csv" "gs://cloudml-public/census/data/adult.test.csv"
 $ flag_estimator_embedding_size: int  8 8 8 8
 $ flag_estimator_hidden_units  :List of 4
  ..$ : int 100
  ..$ : int 70
  ..$ : int 50
  ..$ : int 25
 $ flag_eval_num_epochs         : int  5 5 5 5
 $ flag_eval_batch_size         : int  40 40 40 40
 $ flag_eval_delay_secs         : int  10 10 10 10
 $ flag_eval_steps              : int  100 100 100 100
 $ flag_train_num_epochs        : int  5 5 5 5
 $ flag_train_batch_size        : int  40 40 40 40
 $ flag_train_steps             : int  10 10 10 10
 $ steps_completed              : int  255 255 255 255
 $ metrics                      : chr  "runs/2017-09-27T19-49-07Z/tfruns.d/metrics.json" "runs/2017-09-27T19-49-07Z/tfruns.d/metrics.json" "runs/2017-09-27T19-49-07Z/tfruns.d/metrics.json" "runs/2017-09-27T19-49-07Z/tfruns.d/metrics.json"
 $ script                       : chr  "train.R" "train.R" "train.R" "train.R"
 $ start                        : num  1.51e+09 1.51e+09 1.51e+09 1.51e+09
 $ end                          : num  1.51e+09 1.51e+09 1.51e+09 1.51e+09
 $ completed                    : logi  TRUE TRUE TRUE TRUE
 $ output                       : chr  "\n> library(tensorflow)\n\n> library(tfestimators)\n\n> source(\"model.R\")\n\n> # read in flags\n> FLAGS <- fl"| __truncated__ "\n> library(tensorflow)\n\n> library(tfestimators)\n\n> source(\"model.R\")\n\n> # read in flags\n> FLAGS <- fl"| __truncated__ "\n> library(tensorflow)\n\n> library(tfestimators)\n\n> source(\"model.R\")\n\n> # read in flags\n> FLAGS <- fl"| __truncated__ "\n> library(tensorflow)\n\n> library(tfestimators)\n\n> source(\"model.R\")\n\n> # read in flags\n> FLAGS <- fl"| __truncated__
 $ source_code                  : chr  "runs/2017-09-27T19-49-07Z/tfruns.d/source.tar.gz" "runs/2017-09-27T19-49-07Z/tfruns.d/source.tar.gz" "runs/2017-09-27T19-49-07Z/tfruns.d/source.tar.gz" "runs/2017-09-27T19-49-07Z/tfruns.d/source.tar.gz"
 $ context                      : chr  "local" "local" "local" "local"
 $ type                         : chr  "training" "training" "training" "training"
```

Note how each entry seems to have been duplicated to length four. Note that the `flag_estimator_hidden_units` was a length-four integer vector, but it's now stored as a list of length 4 instead.

**Comment by jjallaire** [comment]
Yikes, that's because of the way I am combining the fields. I have to run for today, if you want to take a look and see if there is a simple fix feel free to do so, otherwise I'll investigate in the AM.

**Comment by jjallaire** [comment]
In general I didn't expect flag values to ever have more than length 1 as they are intended to just be simple character, boolean, or numeric values. I wonder if the real fix would be to just disallow this entirely? I realize that you ran into this during testing though so may have a legit use case. Let me know what you think.

**Comment by jjallaire** [comment]
I am unable to get the command line parser to even parse non-atomic values for flags. I think for now we should just require length one flags (otherwise there is currently no way to specify them from the command line). I'll make this change (happy to reconsider if you think that it's critical we support flags of > 1 length, but note that if we do this we should also upgrade the command line parser to support a list syntax of some kind.

**Comment by jjallaire** [comment]
Here's my change: https://github.com/rstudio/tfruns/commit/23c914b8875660fbaf4d146177b818c02ae68a20

If this seems like a sensible approach you can go ahead and close the issue.

**Comment by kevinushey** [comment]
We might need to handle it -- note that Python's `argparse` allows for it (with e.g. `nargs='+'`), and some TensorFlow examples do take advantage of this. For example:

https://github.com/GoogleCloudPlatform/cloudml-samples/blob/master/census/tensorflowcore/trainer/task.py#L423-L426

FWIW the approach I took in `tensorflow` was to parse arguments as YAML, so that the following would work:

```R
> tensorflow:::parse_arguments("--foo=[1,2,3]")
$foo
[1] 1 2 3
```

but this of course does diverge from how `argparse` handles things, so if we did want to handle it we might want to handle it in the same way as `argparse`, where flags expecting multiple values are just space-separated (e.g. the above would be `--foo 1 2 3`).

I can't recall if I chose YAML-style parsing out of convenience or to mimic some other behavior I saw in a TensorFlow example now :-/

That said, there's nothing stopping us from saying 'flags should be length one'; but users could write something like `--foo="1,2,3"` and the client-side code could manually parse and split the flag value.

**Comment by jjallaire** [comment]
The code you wrote in `tensorflow` was migrated into `tfruns:::parse_flags` so we still have the behavior you outline above.

I think we chose YAML out of convenience. Note that there is nothing stopping us from using argparse directly here as `training_run` will always be used in a context where Python is available.

Let me see if I can fix the issue with the record composition so that if we do introduce support for multiple arguments then this problem doesn't crop back up.

**Comment by jjallaire** [comment]
The problem ends up being in `tibble::as_data_frame` (which we use so that we can cleanly `rbind` multiple run records together):

<img width="1434" alt="screen shot 2017-09-28 at 6 11 58 am" src="https://user-images.githubusercontent.com/104391/30961446-3dcfa3e4-a414-11e7-92e8-24ee4d41fe40.png">

Note that the docs for `tibble::as_data_frame` require that all elements of the list be of the same length.

I tried calling `as.data.frame` instead and I get slightly better behavior however we still get multiple indexed instances of the flag:

<img width="1435" alt="screen shot 2017-09-28 at 6 16 54 am" src="https://user-images.githubusercontent.com/104391/30961616-c14bc81a-a414-11e7-80f9-01d5e335645f.png">

I could convert flags with length > 0 to a deparsed representation? For the purposes of the data frame that would make them character vectors though. That would be fine for display however if people are attempting to compute on them it might cause problems.

Let me know how you think we should proceed or if there is another path I haven't yet seen.

**Comment by kevinushey** [comment]
I'm not a fan of deparsing either. I can't think of a nice solution that would still fit in a `data.frame` :-/

The closest thing I can think of that could work would be to stuff all of the flags into a length-one list, and have that item itself be a list. E.g. (sample code)

```
library(tibble)
FLAGS <- list(a = 1, b = 2, c = 3)
df <- data_frame(run_dir = "<...>")
df$flags <- list(FLAGS)
```

This implies one extra de-reference to pull the flags out, though.

**Comment by kevinushey** [comment]
Maybe we should just punt on this and wait to see if any users in the wild come back asking us to support this? I didn't see much of this kind of usage in the wild so perhaps it's just not worth the effort at this point?

**Comment by jjallaire** [comment]
We want the flags to be individual values on the data frame so that it's
possible to subset on them.

I agree that we should just keep this disabled and then react if need be.
